core/src/main/java/io/grpc/internal/InternalSubchannel.java

```diff
@@ -80,52 +78,36 @@
   private final ChannelTracer channelTracer;
   private final ChannelLogger channelLogger;
 
-  // File-specific convention: methods without GuardedBy("lock") MUST NOT be called under the lock.
-  private final Object lock = new Object();
-
-  // File-specific convention:
-  //
-  // 1. In a method without GuardedBy("lock"), executeLater() MUST be followed by a drain() later in
-  // the same method.
-  //
-  // 2. drain() MUST NOT be called under "lock".
-  //
-  // 3. Every synchronized("lock") must be inside a try-finally which calls drain() in "finally".
+  /**
+   * All field must be mutated in the syncContext.
+   */
   private final SynchronizationContext syncContext;
 
   /**
    * The index of the address corresponding to pendingTransport/activeTransport, or at beginning if
    * both are null.
    */
-  @GuardedBy("lock")
-  private Index addressIndex;
+  private volatile Index addressIndex;
```

Comment:
Why is this volatile? It is never changed. It seems there are reads from the index without synchronization: tostring, getAddressGroups (just missing throwIfNotInThisSynchronizationContext()?). That needs to be resolved.

Reply:
@zhangkun83 and I discussed this before, this was tricky one iirc. Those two methods are called from elsewhere (getAddressGroups is called from ManagedChannelImpl.SubchannelImpl and OobChannel). if it returns Future, we need to create a thread otherwise it will hang. Slightly delay for the getAddressGroups looked acceptable according to Kun. @zhangkun83, can you explain again?

Reply:
My fault. Making `addressIndex` volatile doesn't help. It should in fact be final.
`Index.addressGroups` can be made volatile to help `toString()` and `getAddressGroups()`. They can't have `throwIfNotInThisSynchronizationContext()` because they didn't have such requirement when they were first introduced.

Reply:
`Index.addressGroups` being volatile would fix it, although may be a bit confusing to see all the unnecessary volatile reads (making you think synchronization is going on). We could also keep another field that the same as `Index.addressGroups` except being volatile. In can be in InternalSubchannel (vs Index) if we'd like.

Reply:
@zhangkun83 that's definitely worth a comment.

Reply:
+1 for having a `volatile addressGroups` in `InternalSubchannel`.

Reply:
yeah just added the volatile accessor for `addressGroups`, PTAL.